Thread safety in Observables visitAdditionalChildren methods

maraisr opened the pull request:
#### af27e9ed9ef2f1e26eea8a4541f36ff7c39cb6d0
<pre>
Thread safety in Observables visitAdditionalChildren methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=284500">https://bugs.webkit.org/show_bug.cgi?id=284500</a>

Reviewed by Ryosuke Niwa and Chris Dumez.

visitAdditionalChildren should not ref/deref objects that are
single-threaded ref counted since this function gets called
concurrently to the main thread.

* Source/WebCore/dom/InternalObserverDrop.cpp:
* Source/WebCore/dom/InternalObserverEvery.cpp:
* Source/WebCore/dom/InternalObserverFilter.cpp:
* Source/WebCore/dom/InternalObserverFind.cpp:
* Source/WebCore/dom/InternalObserverFirst.cpp:
* Source/WebCore/dom/InternalObserverForEach.cpp:
* Source/WebCore/dom/InternalObserverInspect.cpp:
* Source/WebCore/dom/InternalObserverLast.cpp:
* Source/WebCore/dom/InternalObserverReduce.cpp:
* Source/WebCore/dom/InternalObserverSome.cpp:
* Source/WebCore/dom/InternalObserverTake.cpp:

Canonical link: <a href="https://commits.webkit.org/287733@main">https://commits.webkit.org/287733@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/adac4d3bd85e5e1f38907323630558767b5ee018

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59655 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34582 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85172 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31630 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82759 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7963 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62987 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20780 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53117 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73416 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43292 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50400 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27576 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30089 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86605 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71279 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70528 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17575 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14540 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13490 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7836 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13355 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7675 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11194 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -222,8 +222,8 @@ class InternalObserverInspect final : public InternalObserver {
 
-    Ref<Subscriber> m_subscriber;
-    ObservableInspector m_inspector;
+    const Ref<Subscriber> m_subscriber;
```
visitAdditionalChildren in this class is also problematic. It ref's a bunch of main-thread only objects.

maraisr replied:
How do I find out if something I'm Refing is main-thread only? And what is my strategy here? 

Should we revert the original pr that introduced this operator?

rniwa replied:
Most of WebCore objects are main-thread only or otherwise thread specific. `SubscriptionObserverCallback` for example inherits from `RefCounted` so it's clearly not multi-thread safe.

maraisr replied:
Are you referring to the `const ObservableInspector m_inspector` here, or what is triggering you to say this class is problematic?

rniwa replied:
See https://github.com/WebKit/WebKit/blob/6c9bf5a79e0ae3933d90140445b19047b9a2bd46/Source/WebCore/dom/InternalObserverInspect.cpp#L176-L189

It's problematic that we're ref'ing `m_inspector.next`, `m_inspector.error`, etc... since those are not multi-thread-safe objects.

maraisr replied:
Would you want to see:
```diff
- if (RefPtr next = m_inspector.next) 
-   next->visitJSFunction(visitor); 
+ m_inspector.next->visitJSFunction(visitor); 
```

I think what I'm understanding is that I can store a Ref as a data member, just "ref" it in this `visitAdditionalChildren` method? As to say, no `protectedXYZ` and no `Ref { m_thing }`

rniwa replied:
Well, we probably need to do a null check so more like:
```cpp
if (m_inspector.next)
    SUPPRESS_UNCOUNTED_ARG m_inspector.next->visitJSFunction(visitor);
```

maraisr replied:
@cdumez do we want to hold on merging this, so I can make this class even less problematic as discussed in this thread? Or should I submit a new patch for it?

rniwa replied:
Submitting a new PR is fine although I don't mind if you pushed a new change here either.

maraisr replied:
@rniwa Is it that we need `SUPPRESS_UNCOUNTED_ARG` here—and not the others—because its a member of a struct, than a `const` here?

maraisr replied:
@rniwa Would I get in trouble if i was to remove the `safe-merge-queue` to push a change?

rniwa replied:
So here, we can't rely on const'ness of a member variable because we're retrieving a value out of `m_inspector`. There is nothing which guarantees that m_inspector's member variables won't change its value over time. I guess in theory, if we made `m_inspector` `const`, we could make the static analyzer smart enough to recognize that its member's values won't change but we don't have such an elaborate const analysis at the moment.

rniwa replied:
@maraisr : no, you won't get in trouble if you removed the label. It just means that someone has to re-add the merge-queue label in order for the PR to get merged.

maraisr replied:
I did just push a change to this, please let me know if this is all good now @rniwa again so sorry about this. Ill take future patches more slowly, and try and spot these earlier on.

maraisr committed:
Thread safety in Observables visitAdditionalChildren methods
https://bugs.webkit.org/show_bug.cgi?id=284500

Reviewed by Ryosuke Niwa and Chris Dumez.

visitAdditionalChildren should not ref/deref objects that are
single-threaded ref counted since this function gets called
concurrently to the main thread.

* Source/WebCore/dom/InternalObserverDrop.cpp:
* Source/WebCore/dom/InternalObserverEvery.cpp:
* Source/WebCore/dom/InternalObserverFilter.cpp:
* Source/WebCore/dom/InternalObserverFind.cpp:
* Source/WebCore/dom/InternalObserverFirst.cpp:
* Source/WebCore/dom/InternalObserverForEach.cpp:
* Source/WebCore/dom/InternalObserverInspect.cpp:
* Source/WebCore/dom/InternalObserverLast.cpp:
* Source/WebCore/dom/InternalObserverReduce.cpp:
* Source/WebCore/dom/InternalObserverSome.cpp:
* Source/WebCore/dom/InternalObserverTake.cpp:

Canonical link: https://commits.webkit.org/287733@main
